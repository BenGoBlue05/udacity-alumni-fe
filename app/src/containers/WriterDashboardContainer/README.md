## WriterDashboardContainer
A container that does ...

### Example Usage

```js
<WriterDashboardContainer  />
```


### Other Information
