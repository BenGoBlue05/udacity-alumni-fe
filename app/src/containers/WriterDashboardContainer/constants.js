export const WRITER_DASHBOARD_SET_ARTICLES = 'WRITER_DASHBOARD_SET_ARTICLES';
export const WRITER_DASHBOARD_SET_ARTICLES_PAGE = 'WRITER_DASHBOARD_SET_ARTICLES_PAGE';
export const WRITER_DASHBOARD_TOGGLE_ASIDE = 'WRITER_DASHBOARD_TOGGLE_ASIDE';
export const WRITER_DASHBOARD_SET_ERROR = 'WRITER_DASHBOARD_SET_ERROR';
export const WRITER_DASHBOARD_CLEAR_ERROR = 'WRITER_DASHBOARD_CLEAR_ERROR';
export const WRITER_DASHBOARD_OPEN_CONFIRMATION_MODAL = 'WRITER_DASHBOARD_OPEN_CONFIRMATION_MODAL';
export const WRITER_DASHBOARD_DELETE_ARTICLE_INITIATION = 'WRITER_DASHBOARD_DELETE_ARTICLE_INITIATION';
export const WRITER_DASHBOARD_DELETE_ARTICLE_SUCCESS = 'WRITER_DASHBOARD_DELETE_ARTICLE_SUCCESS';
export const WRITER_DASHBOARD_DELETE_ARTICLE_FAILURE = 'WRITER_DASHBOARD_DELETE_ARTICLE_FAILURE';
export const WRITER_DASHBOARD_CLEAR_MESSAGE = 'WRITER_DASHBOARD_CLEAR_MESSAGE';
export const WRITER_DASHBOARD_SET_SORT_OPTIONS = 'WRITER_DASHBOARD_SET_SORT_OPTIONS';
