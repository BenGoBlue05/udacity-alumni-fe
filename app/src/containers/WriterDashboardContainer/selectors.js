import { createSelector } from 'reselect';

const getArticles = () => (state) => state.articles.items;
const getArticlesCurrentPage = () => (state) => state.articles.currentPage;
const getArticlesPerPage = () => (state) => state.articles.perPage;

const getArticlesSortIndex = () => (state) => state.articles.sortIndex;
const getArticlesSortAscending = () => (state) => state.articles.sortAscending;

export const getPagedArticles = createSelector(
  getArticles(),
  getArticlesCurrentPage(),
  getArticlesPerPage(),
  (articles, currentPage, perPage) => { // eslint-disable-line
    if (articles && articles.length > 0) {
      const current = currentPage - 1;
      const from = current * perPage;
      const to = current * perPage + perPage;
      return articles.filter((_, i) =>
        i >= from && i < to
      );
    }
  }
);

export const getSortedArticles = createSelector(
  getPagedArticles,
  getArticlesSortIndex(),
  getArticlesSortAscending(),
  (articles, index, isAscending) => { // eslint-disable-line
    if (articles && articles.length > 0) {
      if (index === 1) {
        return articles.sort((a, b) =>
          isAscending ? a.title > b.title : b.title > a.title
        );
      }
      if (index === 2) {
        return articles.sort((a, b) =>
          isAscending ? a.status > b.status : b.status > a.status
        );
      }
      if (index === 3) {
        return articles.sort((a, b) =>
          isAscending ?
            new Date(a.updated_at.split('T')) - new Date(b.updated_at.split('T'))
          :
            new Date(b.updated_at.split('T')) - new Date(a.updated_at.split('T'))
        );
      }
      if (index === 4) {
        return articles.sort((a, b) =>
          isAscending ? a.user.name > b.user.name : b.user.name > a.user.name
        );
      }
      if (index === 5) {
        return articles.sort((a, b) =>
          a.title > b.title
        );
      }
      return articles;
    }
  }
);
