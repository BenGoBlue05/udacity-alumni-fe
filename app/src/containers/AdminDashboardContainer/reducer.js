import * as types from './constants';

export const initialState = {
  activeTab: 0,
  users: {
    editing: null,
    items: [],
    perPage: 10,
    currentPage: 1,
  },
  articles: {
    items: [],
    perPage: 10,
    currentPage: 1,
  },
  aside: {
    isVisible: false,
  },
  error: null,
  userTable: {
    sortIndex: 1,
    sortAscending: true,
  },
  modal: {
    userId: null,
    avatarInput: null,
    isVisible: false,
  },
};

const adminDashboardReducer =
  (state = initialState, action) => {
    switch (action.type) {
      case types.ADMIN_DASHBOARD_OPEN_MODAL:
        return {
          ...state,
          modal: {
            ...state.modal,
            isVisible: true,
            userId: action.userId,
          },
        };
      case types.ADMIN_DASHBOARD_CLOSE_MODAL:
        return {
          ...state,
          modal: {
            ...state.modal,
            isVisible: false,
          },
        };
      case types.ADMIN_DASHBOARD_EDIT_AVATAR_INPUT:
        return {
          ...state,
          modal: {
            ...state.modal,
            avatarInput: action.input,
          },
        };
      case types.ADMIN_DASHBOARD_SET_SORT_OPTIONS:
        return {
          ...state,
          userTable: {
            ...state.userTable,
            sortIndex: action.index,
            sortAscending: action.ascending,
          },
        };
      case types.ADMIN_DASHBOARD_SET_ERROR:
        return {
          ...state,
          error: action.error,
        };
      case types.ADMIN_DASHBOARD_CLEAR_ERROR:
        return {
          ...state,
          error: null,
        };
      case types.ADMIN_DASHBOARD_TOGGLE_ASIDE:
        return {
          ...state,
          aside: {
            isVisible: !state.aside.isVisible,
          },
        };
      case types.ADMIN_DASHBOARD_CLEAR_USER_EDITING:
        return {
          ...state,
          users: {
            ...state.users,
            editing: null,
          },
        };
      case types.ADMIN_DASHBOARD_SET_USER_EDITING:
        return {
          ...state,
          users: {
            ...state.users,
            editing: action.index,
          },
        };
      case types.ADMIN_DASHBOARD_SET_USERS_PAGE:
        return {
          ...state,
          users: {
            ...state.users,
            currentPage: action.page,
          },
        };
      case types.ADMIN_DASHBOARD_SET_ARTICLES_PAGE:
        return {
          ...state,
          articles: {
            ...state.articles,
            currentPage: action.page,
          },
        };
      case types.ADMIN_DASHBOARD_SET_USERS:
        return {
          ...state,
          users: {
            ...state.users,
            items: action.users,
          },
        };
      case types.ADMIN_DASHBOARD_SET_ARTICLES:
        return {
          ...state,
          articles: {
            ...state.articles,
            items: action.articles,
          },
        };
      case types.ADMIN_DASHBOARD_SET_ACTIVE_TAB:
        return {
          ...state,
          activeTab: action.tab,
        };
      default:
        return state;
    }
  };

export default adminDashboardReducer;
