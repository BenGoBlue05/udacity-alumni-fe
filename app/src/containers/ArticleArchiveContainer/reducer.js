import * as types from './constants';
import update from 'react-addons-update';

export const initialState = {
  tag: null,
};

const articleArchiveReducer =
  (state = initialState, action) => {
    switch (action.type) {
      case types.ARCHIVE_SET_TAG:
        return update(state, {
          tag: {
            $set: action.tag,
          },
        });
      default:
        return state;
    }
  };

export default articleArchiveReducer;
