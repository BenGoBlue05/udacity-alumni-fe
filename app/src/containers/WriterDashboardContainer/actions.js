import * as types from './constants';
import { push } from 'react-router-redux';

export const setArticles = (articles) => ({
  type: types.WRITER_DASHBOARD_SET_ARTICLES,
  articles,
});

export const setArticlesPage = (page) => ({
  type: types.WRITER_DASHBOARD_SET_ARTICLES_PAGE,
  page,
});

export const toggleAside = () => ({
  type: types.WRITER_DASHBOARD_TOGGLE_ASIDE,
});

export const setDashboardError = (error) => ({
  type: types.WRITER_DASHBOARD_SET_ERROR,
  error,
});

export const clearDashboardError = () => ({
  type: types.WRITER_DASHBOARD_CLEAR_ERROR,
});

export const setSortOptions = (index, ascending) => ({
  type: types.WRITER_DASHBOARD_SET_SORT_OPTIONS,
  index,
  ascending,
});

export const openConfirmationModal = (id) => ({
  type: types.WRITER_DASHBOARD_OPEN_CONFIRMATION_MODAL,
  id,
});

export const cancelDeletingArticle = () => ({
  type: types.WRITER_DASHBOARD_CANCEL_ARTICLE_DELETION,
});

export const dashboardDeleteArticleInitiation = () => ({
  type: types.WRITER_DASHBOARD_DELETE_ARTICLE_INITIATION,
});

export const dashboardDeleteArticleSuccess = (message) => ({
  type: types.WRITER_DASHBOARD_DELETE_ARTICLE_SUCCESS,
  message,
});

export const dashboardDeleteArticleFailure = (error) => ({
  type: types.WRITER_DASHBOARD_DELETE_ARTICLE_FAILURE,
  error,
});

export const clearDashboardMessage = () => ({
  type: types.WRITER_DASHBOARD_CLEAR_MESSAGE,
});

export const editArticle = (id) => (dispatch) => {
  dispatch(
    push(`/author/cms?action=edit&articleId=${id}`)
  );
};

export const viewArticle = (slug) => (dispatch) => {
  dispatch(
    push(`/articles/${slug}`)
  );
};
