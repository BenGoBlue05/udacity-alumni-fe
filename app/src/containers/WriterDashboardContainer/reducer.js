import * as types from './constants';

export const initialState = {
  articles: {
    items: [],
    perPage: 8,
    currentPage: 1,
    sortIndex: 1,
    sortAscending: true,
  },
  aside: {
    isVisible: false,
  },
  error: null,
  modal: {
    user: null,
    avatarInput: null,
    isVisible: false,
  },
  confirmationModal: {
    isVisible: false,
    articleId: null,
  },
  isSubmitting: false,
  message: null,
};


const writerDashboardReducer =
  (state = initialState, action) => {
    switch (action.type) {
      case types.WRITER_DASHBOARD_SET_ARTICLES:
        return {
          ...state,
          articles: {
            ...state.articles,
            items: action.articles,
          },
        };
      case types.WRITER_DASHBOARD_SET_ARTICLES_PAGE:
        return {
          ...state,
          articles: {
            ...state.articles,
            currentPage: action.page,
          },
        };
      case types.WRITER_DASHBOARD_TOGGLE_ASIDE:
        return {
          ...state,
          aside: {
            isVisible: !state.aside.isVisible,
          },
        };
      case types.WRITER_DASHBOARD_SET_ERROR:
        return {
          ...state,
          error: action.error,
        };
      case types.WRITER_DASHBOARD_CLEAR_ERROR:
        return {
          ...state,
          error: null,
        };
      case types.WRITER_DASHBOARD_OPEN_CONFIRMATION_MODAL:
        return {
          ...state,
          confirmationModal: {
            isVisible: true,
            articleId: action.id,
          },
        };
      case types.WRITER_DASHBOARD_DELETE_ARTICLE_INITIATION:
        return {
          ...state,
          isSubmitting: true,
        };
      case types.WRITER_DASHBOARD_DELETE_ARTICLE_SUCCESS:
        return {
          ...state,
          isSubmitting: false,
          message: action.message,
          confirmationModal: {
            isVisible: false,
            articleId: null,
          },
        };
      case types.WRITER_DASHBOARD_DELETE_ARTICLE_FAILURE:
        return {
          ...state,
          isSubmitting: false,
          error: action.error,
        };
      case types.WRITER_DASHBOARD_CLEAR_MESSAGE:
        return {
          ...state,
          message: null,
        };
      case types.WRITER_DASHBOARD_SET_SORT_OPTIONS:
        return {
          ...state,
          articles: {
            ...state.articles,
            sortIndex: action.index,
            sortAscending: action.ascending,
          },
        };
      default:
        return state;
    }
  };

export default writerDashboardReducer;
