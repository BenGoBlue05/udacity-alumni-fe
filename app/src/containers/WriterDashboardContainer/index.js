import React, { Component, PropTypes } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import * as WriterDashboardActionCreators from './actions';
import { graphql } from 'react-apollo';
import gql from 'graphql-tag';
import Section from 'grommet-udacity/components/Section';
import Heading from 'grommet-udacity/components/Heading';
import Box from 'grommet-udacity/components/Box';
import ExpandIcon from 'grommet-udacity/components/icons/base/Expand';
import ContractIcon from 'grommet-udacity/components/icons/base/Contract';
import Button from 'grommet-udacity/components/Button';
import { FullSection, MainContent, AsideButtonContainer } from './styles';
import { getSortedArticles } from './selectors';
import {
  LoadingIndicator,
  DashboardTable,
  MainAside,
  ToastMessage,
  Divider,
  ConfirmationModal,
} from 'components';

class WriterDashboardContainer extends Component {
  componentWillReceiveProps({ articles }) {
    if (articles && articles !== this.props.articles) {
      this.props.actions.setArticles(articles);
    }
  }
  render() {
    const {
      isLoading,
      articlesConfig,
      errorLoading,
      actions,
      articles,
      isMobile,
      message,
      pagedArticles,
      showAside,
      user,
      confirmationModal,
    } = this.props;
    return (
      <Box
        alignContent="center"
        fill="horizontal"
        align="center"
      >
       {isLoading || !articles ?
         <Section
           align="center"
           justify="center"
           className="loading__box"
         >
           <LoadingIndicator isLoading />
         </Section>
        :
          <FullSection primary direction="row">
            <MainContent
              align="center"
              justify="start"
              pad={{ vertical: 'large' }}
            >
              <Heading align="center">
                Dashboard
              </Heading>
              <Divider />
              <Box>
                {pagedArticles && articles &&
                  <DashboardTable
                    items={pagedArticles}
                    isMobile={isMobile}
                    perPage={articlesConfig.perPage}
                    currentPage={articlesConfig.currentPage}
                    onChangePage={actions.setArticlesPage}
                    allItems={articles}
                    onSort={actions.setSortOptions}
                    sortIndex={articlesConfig.sortIndex}
                    sortAscending={articlesConfig.sortAscending}
                    onDelete={({ id }) => actions.openConfirmationModal(id)}
                    onEdit={({ id }) => actions.editArticle(id)}
                    onShow={({ slug }) => actions.viewArticle(slug)}
                  />
                }
              </Box>
              <AsideButtonContainer>
                <Button
                  plain
                  label={`${showAside ? 'Hide' : 'Show'} Aside`}
                  icon={!showAside ? <ExpandIcon /> : <ContractIcon />}
                  onClick={actions.toggleAside}
                />
              </AsideButtonContainer>
            </MainContent>
            {showAside && user && <MainAside user={user} />}
          </FullSection>
        }
        {errorLoading &&
          <ToastMessage
            message={errorLoading.message}
            onClose={actions.clearDashboardError}
            status="critical"
          />
        }
        {message &&
          <ToastMessage
            message={message}
            onClose={actions.clearDashboardMessage}
            status="ok"
          />
        }
        <ConfirmationModal
          isVisible={confirmationModal.isVisible}
          onConfirm={() => this.handleDeletingArticle()}
          onCancel={() => actions.cancelDeletingArticle()}
          title="Confirm Deletion"
        />
      </Box>
    );
  }
}

WriterDashboardContainer.propTypes = {
  articlesConfig: PropTypes.object.isRequired,
  isLoading: PropTypes.bool.isRequired,
  errorLoading: PropTypes.object,
  articles: PropTypes.array,
  pagedArticles: PropTypes.array,
  actions: PropTypes.object.isRequired,
  message: PropTypes.string,
  isMobile: PropTypes.bool.isRequired,
  confirmationModal: PropTypes.object.isRequired,
  showAside: PropTypes.bool.isRequired,
  user: PropTypes.object.isRequired,
};

const mapStateToProps = (state) => ({
  user: state.app.user,
  authToken: state.app.authToken,
  message: state.writerDashboard.message,
  articlesConfig: state.writerDashboard.articles,
  confirmationModal: state.writerDashboard.confirmationModal,
  showAside: state.writerDashboard.aside.isVisible,
  pagedArticles: getSortedArticles(state.writerDashboard),
  isMobile: state.app.isMobile,
});

// mapDispatchToProps :: Dispatch -> {Action}
const mapDispatchToProps = (dispatch) => ({
  actions: bindActionCreators(
    WriterDashboardActionCreators,
    dispatch
  ),
});

const Container = WriterDashboardContainer;

const fetchArticlesQuery = gql`
  query fetchArticles($authToken: String!) {
    allArticles(auth_token: $authToken) {
      id
      title
      status
      slug
      content
      user {
        name
      }
      image: feature_image
      created_at
      updated_at
      featured
      tags {
        id
        tag
      }
    }
  }
`;

const ContainerWithData = graphql(fetchArticlesQuery, {
  options: (ownProps) => ({
    skip: !ownProps.authToken,
    variables: {
      authToken: ownProps.authToken,
    },
  }),
  props: ({ data: { loading, error, allArticles } }) => ({
    articles: allArticles,
    isLoading: loading,
    errorLoading: error,
  }),
})(Container);

const deleteArticleMutation = gql`
  mutation deleteArticle($authToken: String!, $id: ID!) {
    DeleteArticle(input: { id: $id, auth_token: $authToken }) {
      id: deleted_id
    }
  }
`;

const ContainerWithMutation = graphql(deleteArticleMutation, {
  props: ({ mutate }) => ({
    deleteArticleMutation: mutate,
  }),
})(ContainerWithData);

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ContainerWithMutation);
