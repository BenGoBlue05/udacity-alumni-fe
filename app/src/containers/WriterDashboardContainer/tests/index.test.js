import WriterDashboardContainer from '../index';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import React from 'react';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { initialState as writerDashboard } from '../reducer';
import { initialState as app } from '../../AppContainer/reducer';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

const articles = [{
  id: 4,
  title: 'Current procedural terminology Benefit Impaired risk',
  featured: false,
  spotlighted: false,
  content: 'Diagnostic related group HIPAA accreditation',
  feature_image: 'https://robohash.org/sapienteeaoptio.png?size=300x300',
  status: 'published',
  slug: 'current-procedural-terminology-benefit-impaired-risk',
  user: {
    id: 3,
    email: 'andreas@gmail.com',
    created_at: '2016-09-30T02:39:57.297Z',
    updated_at: '2016-10-03T18:49:42.965Z',
    name: 'Andreas Daimainger',
    avatar: 'https://avatars3.githubusercontent.com/u/13679375?v=3&s=466',
    bio: 'Hello world',
    role: 'user',
  },
  created_at: '2016-09-30T02:39:57.352Z',
  updated_at: '2016-09-30T02:39:57.352Z',
}];

const error = {message: 'error in loading, go get a drink of water for a break'};

describe('<WriterDashboard />', () => {
  it('should render with default props', () => {
    const store = mockStore({ writerDashboard, app });
    const wrapper = shallow(
      <WriterDashboardContainer
      	store={store} 
       	isLoading={false}
  		errorLoading={error}
  		articles={articles}
  	  />
    );
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
});
