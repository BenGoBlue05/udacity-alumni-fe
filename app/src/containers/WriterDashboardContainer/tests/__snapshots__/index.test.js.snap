exports[`<WriterDashboard /> should render with default props 1`] = `
<Apollo(Apollo(WriterDashboardContainer))
  actions={
    Object {
      "cancelDeletingArticle": [Function anonymous],
      "clearDashboardError": [Function anonymous],
      "clearDashboardMessage": [Function anonymous],
      "dashboardDeleteArticleFailure": [Function anonymous],
      "dashboardDeleteArticleInitiation": [Function anonymous],
      "dashboardDeleteArticleSuccess": [Function anonymous],
      "editArticle": [Function anonymous],
      "openConfirmationModal": [Function anonymous],
      "setArticles": [Function anonymous],
      "setArticlesPage": [Function anonymous],
      "setDashboardError": [Function anonymous],
      "setSortOptions": [Function anonymous],
      "toggleAside": [Function anonymous],
      "viewArticle": [Function anonymous]
    }
  }
  articles={
    Array [
      Object {
        "content": "Diagnostic related group HIPAA accreditation",
        "created_at": "2016-09-30T02:39:57.352Z",
        "feature_image": "https://robohash.org/sapienteeaoptio.png?size=300x300",
        "featured": false,
        "id": 4,
        "slug": "current-procedural-terminology-benefit-impaired-risk",
        "spotlighted": false,
        "status": "published",
        "title": "Current procedural terminology Benefit Impaired risk",
        "updated_at": "2016-09-30T02:39:57.352Z",
        "user": Object {
          "avatar": "https://avatars3.githubusercontent.com/u/13679375?v=3&s=466",
          "bio": "Hello world",
          "created_at": "2016-09-30T02:39:57.297Z",
          "email": "andreas@gmail.com",
          "id": 3,
          "name": "Andreas Daimainger",
          "role": "user",
          "updated_at": "2016-10-03T18:49:42.965Z"
        }
      }
    ]
  }
  articlesConfig={
    Object {
      "currentPage": 1,
      "items": Array [],
      "perPage": 8,
      "sortAscending": true,
      "sortIndex": 1
    }
  }
  authToken={null}
  confirmationModal={
    Object {
      "articleId": null,
      "isVisible": false
    }
  }
  errorLoading={
    Object {
      "message": "error in loading, go get a drink of water for a break"
    }
  }
  isLoading={false}
  isMobile={false}
  message={null}
  showAside={false}
  store={
    Object {
      "clearActions": [Function clearActions],
      "dispatch": [Function anonymous],
      "getActions": [Function getActions],
      "getState": [Function getState],
      "replaceReducer": [Function replaceReducer],
      "subscribe": [Function subscribe]
    }
  }
  user={null} />
`;
